src/AutoRest.CSharp/Mgmt/Generation/MgmtLongRunningOperationWriter.cs

```diff
@@ -2,157 +2,127 @@
 {
-    internal class MgmtLongRunningOperationWriter : LongRunningOperationWriter
+    internal class MgmtLongRunningOperationWriter
     {
-        private const string _armClientField = "_armClient";
-
-        protected override CSharpType GetBaseType(LongRunningOperation operation)
-        {
-            MgmtLongRunningOperation mgmtOperation = AsMgmtOperation(operation);
-            return mgmtOperation.WrapperResource != null ? new CSharpType(typeof(Operation<>), mgmtOperation.WrapperResource.Type) : base.GetBaseType(operation);
-        }
-
-        protected override CSharpType? GetInterfaceType(LongRunningOperation operation)
-        {
-            MgmtLongRunningOperation mgmtOperation = AsMgmtOperation(operation);
-            return mgmtOperation.WrapperResource != null ? new CSharpType(typeof(IOperationSource<>), mgmtOperation.WrapperResource.Type) : base.GetInterfaceType(operation);
-        }
-
-        protected override CSharpType GetHelperType(LongRunningOperation operation)
+        private readonly CodeWriter _writer;
+        private readonly BuildContext<MgmtOutputLibrary> _context;
+        private readonly string _name;
+        private readonly string _genericString;
+        private readonly bool _isGeneric;
+        private readonly string _waitMethod;
+        private readonly Type _operationType;
+        private readonly Type _responseType;
+        private readonly Type _operationOrResponseType;
+        private readonly FormattableString _operationSourceString;
+        private readonly string _sourceString;
+
+        public string Filename { get; }
+
+        public MgmtLongRunningOperationWriter(BuildContext<MgmtOutputLibrary> context, bool isGeneric)
         {
-            MgmtLongRunningOperation mgmtOperation = AsMgmtOperation(operation);
-            return mgmtOperation.WrapperResource != null ? new CSharpType(typeof(OperationInternals<>), mgmtOperation.WrapperResource.Type) : base.GetHelperType(operation);
+            _writer = new CodeWriter();
+            _context = context;
+            _isGeneric = isGeneric;
+            _name = $"{context.DefaultNamespace.Split('.').Last()}ArmOperation";
+            _genericString = isGeneric ? "<T>" : string.Empty;
+            Filename = isGeneric ? $"LongRunningOperation/{_name}OfT.cs" : $"LongRunningOperation/{_name}.cs";
+            _waitMethod = isGeneric ? "WaitForCompletionAsync" : "WaitForCompletionResponseAsync";
+            _operationType = isGeneric ? typeof(ArmOperation<>) : typeof(ArmOperation);
+            _responseType = isGeneric ? typeof(Response<>) : typeof(Response);
+            _operationOrResponseType = isGeneric ? typeof(OperationOrResponseInternals<>) : typeof(OperationOrResponseInternals);
+            _operationSourceString = isGeneric ? (FormattableString)$"{typeof(IOperationSource<>)} source, " : (FormattableString)$"";
+            _sourceString = isGeneric ? "source, " : string.Empty;
         }
```

Comment:
After the refactor, our writer should only do the writing and do not calculate anything. Therefore should we also introduce a `TypeProvider` to do the calculations and make this writer only writes things?

Reply:
Since this is a static class meaning the definition will never change I think its ok.  Type providers make it nice for variable signatures / methods / ctors / etc

Reply:
Emmm, did not get the point. Neither is this class a static class, nor the class that this writer is writing. Could you please be more specific?

Reply:
The definition is static.  The methods, ctors, signatures will never change.  Unlike a resource or restOperations etc where the methods are constantly different.

Reply:
OK, make sense. But I still think we should introduce a `TypeProvider` for this, for consistency. And also we will do not have to concatenating the name of this class again like here: https://github.com/Azure/autorest.csharp/blob/b07f41b5b41e1f2fceef7ca4b0b638aaf0ff1d91/src/AutoRest.CSharp/Mgmt/Generation/MgmtClientBaseWriter.cs#L68

Reply:
Lets make a future work item for this post GA given our tight timelines right now